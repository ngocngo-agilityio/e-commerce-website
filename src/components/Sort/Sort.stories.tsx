import type { Meta, StoryObj } from '@storybook/react';

// Components
import Sort from '.';

// Constants
import { SORT_OPTIONS } from '@constants';

const meta = {
  title: 'Sort',
  component: Sort,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Sort>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    options: SORT_OPTIONS,
    defaultValue: SORT_OPTIONS[0].label,
    onChange: () => {},
  },
};
