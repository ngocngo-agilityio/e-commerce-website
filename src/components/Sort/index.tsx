'use client';

// Libs
import { memo, useState, MouseEvent, useCallback, useMemo } from 'react';
import {
  Input,
  InputGroup,
  InputRightElement,
  Menu,
  MenuButton,
  MenuItem,
  MenuList,
  Radio,
  RadioGroup,
} from '@chakra-ui/react';
import isEqual from 'react-fast-compare';

// Assets
import { FilterIcon } from '@assets';

// Types
import { SortOption } from '@types';

interface Props {
  options: SortOption[];
  defaultValue?: string;
  placeholder?: string;
  onChange: (value: string) => void;
}

const Sort = ({
  options = [],
  defaultValue,
  placeholder = 'Sort by',
  onChange,
}: Props): JSX.Element => {
  const [selectedOption, setSelectedOption] = useState(defaultValue);

  const handleOnChange = useCallback(
    (e: MouseEvent<HTMLButtonElement>) => {
      const value = e.currentTarget.value;

      setSelectedOption(value);
      onChange(value);
    },
    [onChange],
  );

  const getLabelByValue = useMemo(() => {
    const option = options.find((option) => option.value === selectedOption);

    return option ? option.label : '';
  }, [options, selectedOption]);

  return (
    <Menu closeOnSelect={false} size="base">
      <MenuButton>
        <InputGroup w={{ base: 'full', md: '245px' }}>
          <Input
            variant="outline"
            size="base"
            borderRadius="base"
            placeholder={placeholder}
            value={getLabelByValue}
          />
          <InputRightElement h="24px" my="12px" mr="7px">
            <FilterIcon />
          </InputRightElement>
        </InputGroup>
      </MenuButton>

      <MenuList>
        <RadioGroup value={selectedOption}>
          {options.map((item) => {
            const { value = '', label = '' } = item || {};

            return (
              <MenuItem key={value} value={value} onClick={handleOnChange}>
                <Radio value={value} colorScheme="orange">
                  {label}
                </Radio>
              </MenuItem>
            );
          })}
        </RadioGroup>
      </MenuList>
    </Menu>
  );
};

export default memo(Sort, isEqual);
