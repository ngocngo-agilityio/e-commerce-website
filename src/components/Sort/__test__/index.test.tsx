// Libs
import { fireEvent, render, waitFor } from '@testing-library/react';

// Mocks
import { MOCK_SORT_OPTIONS } from '@mocks';

// Types
import { SortOption } from '@types';

// Components
import Sort from '..';

const mockProps = {
  options: MOCK_SORT_OPTIONS,
  onChange: jest.fn(),
};

describe('Sort component', () => {
  test('should match snapshot for Sort', () => {
    const { container } = render(<Sort {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  test('should call onChange when select option', () => {
    const mockOptions = MOCK_SORT_OPTIONS;
    mockOptions[2] = null as unknown as SortOption;

    const { getByText } = render(
      <Sort {...mockProps} defaultValue={MOCK_SORT_OPTIONS[0].value} />,
    );

    fireEvent.click(getByText(MOCK_SORT_OPTIONS[1].label));

    waitFor(() =>
      expect(mockProps.onChange).toHaveBeenCalledWith(
        MOCK_SORT_OPTIONS[1].value,
      ),
    );
  });
});
