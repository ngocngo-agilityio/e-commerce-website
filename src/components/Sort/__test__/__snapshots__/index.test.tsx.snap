// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort component should match snapshot for Sort 1`] = `
<div>
  <button
    aria-controls="menu-list-:r8:"
    aria-expanded="false"
    aria-haspopup="menu"
    class="chakra-menu__menu-button css-1eqi7qs"
    id="menu-button-:r8:"
  >
    <span
      class="css-xl71ch"
    >
      <div
        class="chakra-input__group css-1brjfbh"
        data-group="true"
      >
        <input
          class="chakra-input css-1lqcdac"
          placeholder="Sort by"
          value=""
        />
        <div
          class="chakra-input__right-element css-8p9vs9"
        >
          <svg
            class="chakra-icon css-g7h3az"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 4L10 4"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M14 4L21 4"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M3 12L12 12"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M16 12L21 12"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M3 20L8 20"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M12 20L21 20"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M10 1L10 7"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M16 9L16 15"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M8 17L8 23"
              stroke="#B7BFC5"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </span>
  </button>
  <div
    class="css-0"
    style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
  >
    <div
      aria-orientation="vertical"
      class="chakra-menu__menu-list css-sjt5nk"
      id="menu-list-:r8:"
      role="menu"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8);"
      tabindex="-1"
    >
      <div
        class="chakra-radio-group css-0"
        role="radiogroup"
      >
        <button
          class="chakra-menu__menuitem css-1k756t9"
          data-index="0"
          id="menu-list-:r8:-menuitem-:ra:"
          role="menuitem"
          tabindex="-1"
          type="button"
          value="1"
        >
          <label
            class="chakra-radio css-1nqw95c"
          >
            <input
              class="chakra-radio__input"
              id="radio-:rb:"
              name="radio-:r9:"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="radio"
              value="1"
            />
            <span
              aria-hidden="true"
              class="chakra-radio__control css-ssalds"
            />
            <span
              class="chakra-radio__label css-1y8kf23"
            >
              Increment by name
            </span>
          </label>
        </button>
        <button
          class="chakra-menu__menuitem css-1k756t9"
          data-index="1"
          id="menu-list-:r8:-menuitem-:rc:"
          role="menuitem"
          tabindex="-1"
          type="button"
          value="2"
        >
          <label
            class="chakra-radio css-1nqw95c"
          >
            <input
              class="chakra-radio__input"
              id="radio-:rd:"
              name="radio-:r9:"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="radio"
              value="2"
            />
            <span
              aria-hidden="true"
              class="chakra-radio__control css-ssalds"
            />
            <span
              class="chakra-radio__label css-1y8kf23"
            >
              Decrement by name
            </span>
          </label>
        </button>
      </div>
    </div>
  </div>
</div>
`;
